### PR TITLE
Nested scroll effect

### PR DESCRIPTION
**Pull request by Hixie**
@abarth This works (as demonstrated) but definitely needs more work. In particular:
- some things should be split out into separate better tested and smaller PRs.
- some of the refactoring doesn't really make sense and should be improved.
It also has a few bugs (especially on iOS) though I'm not sure how to reproduce them reliably.

I'm totally happy to mechanically do that work, but I'm burned out on the design side. I suggest that we do a high-level overview review pass first, where you see if there's any obvious chunks to split out, or any obvious improvements to the way this is factored, and then I use your suggestions to split the code down into the smaller PRs and we do more detailed reviews on those.

### CODE REVIEW

**Comment by abarth** [comment]
Just started looking, but there's a lot of obviously good stuff that can be pulled out into a warmup patch to make this diff smaller.  (Will provide more useful comments as I dig in further.)

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 1]
```diff
@@ -0,0 +1,985 @@
...
+  final Widget innerBox;
+
+  double get initialScrollOffset => 0.0;
+
+  @protected
+  List<Widget> buildSlivers(BuildContext context, ScrollController innerController, bool innerBoxIsScrolled) {
+    final List<Widget> slivers = <Widget>[];
+    slivers.addAll(outerSliverBuilder(context, innerBoxIsScrolled));
+    slivers.add(new SliverFillRemaining(
+      child: new PrimaryScrollController(
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: This introduces and interesting vertical/horizontal asymmetry.

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 2]
```diff
@@ -0,0 +1,985 @@
...
+          extra = _outerPosition.pixels - (_outerPosition.maxScrollExtent - _outerPosition.minScrollExtent);
+        }
+        assert(extra <= 0.0);
+        minRange = _outerPosition.minScrollExtent;
+        maxRange = _outerPosition.maxScrollExtent + extra;
+        assert(minRange <= maxRange);
+        correctionOffset = 0.0;
+      }
+    }
+    return new _NestedScrollingSituation(
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: oh man

**Comment by Hixie** [comment]
Ok I've pulled out all the self-contained stuff.

What remains is as follows:
* ScrollActivity getting moved into its own file that doesn't depend on ScrollPosition.
* ScrollPhysics getting moved into its own file that doesn't depend on ScrollPosition.
* ScrollMetrics gettig moved into its own file that doesn't depend on ScrollPosition.
* ScrollPosition getting split into ScrollPosition and ScrollIndependentPosition.
* An implementation of NestedScrollView to demonstrate how this is useful.
* An implementation of LinkedScrollView to show that other models are also possible.
* Fixes required to make things work after the above changes are applied.

My three biggest problems with this patch are (1) the interfaces I extracted into scroll_interfaces feel a bit strange and definitely have weird names, (2) the actual split between ScrollPosition and ScrollIndependentPosition is weird, and (3) there's still some code duplication in NestedScrollView which implies that the refactoring isn't quite right yet.

PTAL. I'm in particular eager to see suggestions for how to improve the factoring.

**Comment by Hixie** [comment]
I have updated the patch as discussed. Let's take a look at it again tomorrow and see what more we can improve.

File: packages/flutter/lib/src/widgets/scroll_metrics.dart [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+/// range. The [atEdge] getter will return true if the [pixels] position equals
+/// either the [minScrollExtent] or the [maxScrollExtent].
+///
+/// The dimensions of the viewport in the given [axis] are described by
+/// [viewportDimension].
+///
+/// The above values are also exposed in terms of [extentBefore],
+/// [extentInside], and [extentAfter], which may be more useful for use cases
+/// such as scroll bars; for example, see [Scrollbar].
+abstract class ScrollMetrics {
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/scroll_metrics.dart]: abstract const constructor

File: packages/flutter/lib/src/widgets/scroll_metrics.dart [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+  @override
+  String toString() {
+    return '$runtimeType(${extentBefore.toStringAsFixed(1)}..[${extentInside.toStringAsFixed(1)}]..${extentAfter.toStringAsFixed(1)}})';
+  }
+}
+
+@immutable
+class FixedScrollMetrics extends ScrollMetrics {
+  FixedScrollMetrics({
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/scroll_metrics.dart]: const

File: packages/flutter/lib/src/widgets/scroll_metrics.dart (hunk 3 above)
Comment (Hixie) [inline packages/flutter/lib/src/widgets/scroll_metrics.dart]: unfortunately, https://github.com/dart-lang/sdk/issues/29395 https://github.com/dart-lang/sdk/issues/29396

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 5]
```diff
@@ -0,0 +1,1002 @@
...
+
+  // TODO(ianh): we should expose a controller so you can call animateTo, etc.
+
+  final Axis scrollDirection;
+
+  final bool reverse;
+
+  final ScrollPhysics physics;
+
+  final NestedScrollViewOuterSliversBuilder outerSliverBuilder;
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: `headerBuilder`

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 6]
```diff
@@ -0,0 +1,1002 @@
...
+
+  final Widget innerBox;
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: `body`

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 7]
```diff
@@ -0,0 +1,1002 @@
...
+
+  double get initialScrollOffset => 0.0;
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: ???

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart (hunk 1 above)
Comment (abarth) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: Needs documentation that horizontally scrolling children need to set primary explicitly to pick this up.

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 8]
```diff
@@ -0,0 +1,1002 @@
...
+  }) : super(key: key) {
+    assert(scrollDirection != null);
+    assert(reverse != null);
+    assert(outerSliverBuilder != null);
+    assert(innerBox != null);
+  }
+
+  // TODO(ianh): we should expose a controller so you can call animateTo, etc.
+
+  final Axis scrollDirection;
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: Maybe force vertical for now.

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 9]
```diff
@@ -0,0 +1,1002 @@
...
+
+  bool get hasScrolledInnerBox {
+    for (_NestedScrollPosition position in _innerPositions) {
+      if (position.pixels > position.minScrollExtent)
+        return true;
+    }
+    return false;
+  }
+
+  ScrollDirection get userScrollDirection => _userScrollDirection;
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: Maybe remove all these public names.

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 10]
```diff
@@ -0,0 +1,1002 @@
...
+      pixels: pixels,
+      viewportDimension: _outerPosition.viewportDimension,
+      axisDirection: _outerPosition.axisDirection,
+      minRange: minRange,
+      maxRange: maxRange,
+      correctionOffset: correctionOffset,
+    );
+  }
+
+  double unnestOffset(double value, _NestedScrollPosition source) {
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: Can these names be private too?

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 11]
```diff
@@ -0,0 +1,1002 @@
...
+      scrollDirection: widget.scrollDirection,
+      reverse: widget.reverse,
+      physics: new ClampingScrollPhysics(parent: widget.physics),
+      controller: _outerController,
+      slivers: widget.buildSlivers(context, _innerController, hasScrolledInnerBox),
+    );
+  }
+}
+
+class _NestedScrollingSituation extends ScrollMetrics {
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: `_NestedScrollMetrics`

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 12]
```diff
@@ -0,0 +1,1002 @@
...
+
+  bool get _reversed {
+    assert(owner?._outerPosition?.axisDirection != null);
+    switch (owner._outerPosition.axisDirection) {
+      case AxisDirection.up:
+      case AxisDirection.left:
+        return true;
+      case AxisDirection.down:
+      case AxisDirection.right:
+        return false;
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: Duplicated in _getRightWayUp, _reverseDirection.  Probably worth breaking out into a separate function.

File: packages/flutter/lib/src/widgets/scroll_interfaces.dart [hunk 13]
```diff
@@ -0,0 +1,18 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/scroll_interfaces.dart]: Rename file to scroll_context.dart

File: packages/flutter/lib/src/widgets/scroll_activity.dart [hunk 14]
```diff
@@ -0,0 +1,357 @@
...
+  }
+
+  @override
+  bool get shouldIgnorePointer => false;
+
+  @override
+  bool get isScrolling => false;
+}
+
+class DragScrollActivity extends ScrollActivity implements Drag {
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/scroll_activity.dart]: Consider not implementing drag to match the nested case better

**Comment by Hixie** [comment]
Closing this since @abarth has taken it over.

**Pull request closed without merging**